Revised PerspectiveCamera.

tschw opened the pull request:
- Allows extra skew for stereo rendering.
- Allows to query effective FOV angle (considering .zoom).
- Proper copy / serialization of all properties.
- Tidier, single code path frustum calculation.
- No longer uses ProjectionMatrix.makePerspective.

See discussion in #7036 with @bhouston, about smooth interaction with tools, realistic DOF, lens settings and a higher level API for stereo rendering.

To verify the refactor in the first commit is non-breaking:
http://ci.threejs.org/api/pullrequests/8495/examples/webgl_multiple_canvases_grid.html
http://ci.threejs.org/api/pullrequests/8495/examples/webgl_multiple_canvases_complex.html

tschw committed:
Revised PerspectiveCamera.

- Allows extra skew for stereo rendering.
- Allows to query effective FOV angle (considering .zoom).
- Proper copy / serialization of all properties.
- Tidier, single code path frustum calculation.
- No longer uses ProjectionMatrix.makePerspective.

bhouston commented:
Maybe we just need a single function called:
```
.setFilm( filmSize, filmOffset ) {
  // set skew as appropriate.
}
```

I am not sure we need to have focalLength as part of the camera, especially since it isn't used by the setLens().  Maybe just have the advanced physical stuff (lens, film back) as a few functions without trying for some type of universal support.  Avoids having to over-systematize things.

bhouston commented:
This is great without any changes.

tschw commented:
@bhouston 
> I am not sure we need to have `focalLength` as part of the camera

Second that. Especially not like it's (not) integrated, right now. Should be a getter, at best.

mrdoob commented on src/cameras/PerspectiveCamera.js:
```diff
@@ -135,6 +151,18 @@ THREE.PerspectiveCamera.prototype.copy = function ( source ) {
+	this.skewSlopeX = source.skewSlopeX;
+	this.skewSlopeY = source.skewSlopeY;
+
+	if ( source.view !== null && source.view !== undefined ) {
```
I think only checking for `null` should be enough?

MasterJames replied:
For the record the non-triple not-equals matches both null and undefined but if you want to check only null it's a moot point.
I'm unclear why not using undefined isn't prefered?  Best guess null is not an object and undefined is so it takes more resources unessecarily? 
!= undefined covers null too.

mrdoob replied:
> I'm unclear why not using undefined isn't prefered?

Because that changes V8 hidden object.

tschw replied:
> I think only checking for `null` should be enough?

It's called by the loader and since it didn't get serialized before might be missing, thus `=== undefined`.

> > I'm unclear why not using `undefined` isn't prefered?

Because there's a conceptual difference between a missing property and an existing property which currently doesn't bind to an object. Also it's lengthy.

> Because that changes V8 hidden object.

So does defining a property with a value of `undefined`, actually (`for...in` must find it, for example). Also, hidden types are not a speciality of V8 - they exist in pretty much all JS engines used by browsers, and (as long as the order in which properties are added is the same) the transitions extending objects are cached in all of them. Extending an object still needs reallocation, because the extended type needs more space to store its extra properties.

mrdoob replied:
>> I think only checking for `null` should be enough?

> It's called by the loader and since it didn't get serialized before might be missing, thus `=== undefined`.

Which loader?

tschw replied:
> Which loader?

Oh great! `ObjectLoader` doesn't support any of the newer properties, although they are present in `toJSON`... 

"The loader" refers to `JSON.parse` then, I guess :-).

mrdoob commented:
> I am not sure we need to have focalLength as part of the camera, especially since it isn't used by the setLens().

`focalLenght` is used by the `StereoCamera`. Whether that's the right approach or not, I don't know 😇

tschw commented:
> > I am not sure we need to have focalLength as part of the camera, especially since it isn't used by the setLens().

> focalLenght is used by the StereoCamera. Whether that's the right approach or not, I don't know :innocent:

In that case, I'm not sure it's appropriately named: When photographers talk about "a so-and-so-much millimeter lens" they are talking about the focal length. When the [film gauge](https://en.wikipedia.org/wiki/Film_gauge) is known, it can be calculated from `fov`.

`StereoCamera` seems to expect a `zeroParallax` distance - which one might say is the "stereo focus", but it applies on the other side of the lens at a different scale.

bhouston commented:
> which one might say is the "stereo focus"

If we wanted a complete camera, ```Camera.focusDistance``` would be something to add -- I did add it to Clara.io's cameras for DOF purposes.

Usually ```focusDistance``` is the same as the ```zeroParallax``` distance in humans because our eyes both align and focus at the same point (see [Accomodation-convergence  reflex](https://en.wikipedia.org/wiki/Accommodation_reflex)), but not coupling these effects is simulating a disease, which may be desired for medical purposes, but not the likely use case.

tschw commented:
@bhouston 
> If we wanted a complete camera, `Camera.focusDistance` would be something to add -- I did add it to Clara.io's cameras for DOF purposes.

OK, let's rename `.focalLength` to `focusDistance`? Or just `focus`?

> Usually `focusDistance` is the same as the `zeroParallax` distance in humans because our eyes both align and focus at the same point (see Accomodation-convergence reflex), but not coupling these effects is simulating a disease, which may be desired for medical purposes, but not the likely use case.

Yeah, maybe it's best we tie the effects that produce depth cues to a single input. This way, everything "just works". When the need arises (e.g. in a medical Three.js app, or a game simulating drunkenness), we can still add properties to offset one against the other.

bhouston commented:
> OK, let's rename .focalLength to focusDistance? Or just focus?

```focus``` seems appropriate, as it follows the pattern established by ```near``` and ```far```.

I think tying zeroParallax and focus together is fine. :)  We could actually use focus on the StereoCamera as the zeroParallax point for the time being.  At some point we could add an API that looks like:

```
StereoCamera.focus = 10; // sets the focus on both eye cameras.
StereoCamera.autoZeroParallax = false; // auto mode makes zeroParallax follow the focus distance.
StereoCamera.zeroParallax = 10;
```

That allows setting it separately from focus when the time comes.

mrdoob commented:
> OK, let's rename `.focalLength` to `focusDistance`? Or just `focus`?

`focus` sounds good to me too!

tschw committed:
Revised PerspectiveCamera physical interface.

tschw commented:
**Second commit summary / rationale:**

New properties:

`.filmGauge`:  This property sets a frame of reference for other photographic / cinematic settings; the focal length of the lens and the (horizontal) film offset, in particular. It describes the size of the film as "written on its package". 
For example, if `.setFocalLength` is supposed to accept a value in millimeters as for a 35mm full frame camera, its value must be 35 (that's the default). It applies to the maximum of image width and height (the image must fit on the film along the longer axis and we don't make assumptions of how "the camera is held", IOW, whether the viewport has portrait or landscape aspect).

This is fairly close to established standards in digital production, but we opted for a scalar instead of a vector. This way, we avoid a lot of esoteric complexity when the aspect changes.

Changed properties:

`.filmOffset`: (introduced as `skewSlopeX` in the previous commit in respect to a film width of 1): This property describes the horizontal offset of the film, creating an off-center projection for stereoscopic rendering.

`.focus`: (previously incorrectly called `focalLength`). Describes the world (*) distance to a focussed object. This property does not have a direct influence on the projection matrix calculated by this class. It is intended to be consumed by `StereoCamera` and postprocessing filters to simulate a depth-of-field effect.

(*) Maybe not yet, but that's how it should be.

Removed properties:

`.skewSlopeY`: Has been there for completeness, but would end up always 0. 

New methods:

`.setFocalLength`: Sets the FOV angle by focal length in respect to the current film gauge.
`.getFocalLength`: Returns the focal length calculated from the current FOV and the current film gauge.
`.getFilmWidth`: Returns the width of the film (aspect applied)
`.getFilmHeight`: Returns the height of the film (aspect applied)

Deprecated methods:

`setLens`: Same as .setFocalLength, also optionally setting .filmGauge (but the second parameter is rarely used and never actually fed an argument). Previously was imprecise, because the user had to apply a fixed aspect. This is no longer the case - the focal length always applies to the current image fit onto the film.

tschw commented:
Maybe we should keep `setLens` and deprecate its second parameter, remove `setFocalLength` and rename `getFocalLength` to `getLens`? Names would be shorter, but maybe less clear? Not sure.

mrdoob commented:
What do you think about keeping `PerspectiveCamera` simple. And add a new camera (in core) called `FilmCamera` with all the new properties/methods. Or maybe refactor `CinematicCamera` and move it to core?

WestLangley commented:
>What do you think about keeping PerspectiveCamera simple. And add a new camera (in core)

`PhysicalCamera`?

Although it would be nice to avoid:
```javascript
camera instanceof THREE.PerspectiveCamera || camera instanceof THREE.PhysicalCamera
```

I think extending `THREE.PerspectiveCamera` would solve that problem.

tschw commented:
@mrdoob 
> What do you think about keeping `PerspectiveCamera` simple. And add a new camera (in core) called `FilmCamera` with all the new properties/methods. 

I'd basically revert the changes of the second commit, and remove `focalLength`, `setLens` and `skewSlopeY`. This would revive `skewSlopeX` (to allow the higher level to inherit the matrix calculation as-is) and also keep the tidier code.

`PerspectiveCamera` will then wrap the projection matrix in a mean and lean fashion and nothing else.

`StereoCamera` would use the new class with `focus` added (also needed to control DOF).

> Or maybe refactor CinematicCamera and move it to core?

I think we should base the DOF effect on the new API and make it "plug & play". 

The `CinematicCamera` in the examples folder will then be empty / redundant and can simply be removed (the DOF-specific calculation would move into the effect).

Makes sense?

bhouston commented:
Just remember we want focus on perspective camera as well.

Best regards,
Ben Houston
http://Clara.io Online 3d modeling and rendering
On Apr 3, 2016 9:51 AM, "tschw" <notifications@github.com> wrote:

> @mrdoob <https://github.com/mrdoob>
>
> What do you think about keeping PerspectiveCamera simple. And add a new
> camera (in core) called FilmCamera with all the new properties/methods.
>
> I'd basically revert the changes of the second commit, and remove
> focalLength, setLens and skewSlopeY. This would revive skewSlopeX (to
> allow the higher level to inherit the matrix calculation as-is) and also
> keep the tidier code.
>
> PerspectiveCamera will then wrap the projection matrix in a mean and lean
> fashion and nothing else.
>
> StereoCamera would use the new class with focus added (also needed to
> control DOF).
>
> Or maybe refactor CinematicCamera and move it to core?
>
> I think we should base the DOF effect on the new API and make it "plug &
> play".
>
> The CinematicCamera in the examples folder will then be empty / redundant
> and can simply be removed (the DOF-specific calculation would move into the
> effect).
>
> Makes sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mrdoob/three.js/pull/8495#issuecomment-204981734>
>

mrdoob commented:
> Makes sense?

Yep!

> I think extending `THREE.PerspectiveCamera` would solve that problem.

Yes, that's what I meant.

tschw commented:
@mrdoob 
> Or maybe refactor CinematicCamera and move it to core?

@bhouston 
> Just remember we want focus on perspective camera as well.

Actually, I think we should move that one to the other class. There's technically no problem, since it's supposed to be in-core, and requiring the use of a `PhysicalCamera` for anything that focusses makes perfect logical sense.

This way, the line is drawn most clearly between a low level controller for the projection matrix and the traits of a physical camera we model. Granted, `focus` does influence projection matrices when inherited by `StereoCamera` - but in that case, the projection matrices that matter are actually the ones of the eye cameras and not the inherited one (in fact, once I get to the stereo calculation, there will be zero impact on the equivalent mono frustum / combined frustum resembled by the two eye cameras).

mrdoob commented:
I hear you. I wonder if there is a way to not force the transition though.

As far as I understand, you're proposing that anyone hoping to do stereography/vr will need to change to a different camera. Correct?

tschw commented:
> As far as I understand, you're proposing that anyone hoping to do stereography/vr will need to change to a different camera. Correct?

`StereoCamera` could promote it automatically.
 
But there also shouldn't be any problem requiring a different constructor, since everything in the extended cam is strictly optional.

> I wonder if there is a way to not force the transition though.

Renaming would do. `PerspectiveCamera` -> `RawPerspectiveCamera` (this one's for shadows and for who don't care) envisioned `PhysicalCamera` -> `PerspectiveCamera` (this one becomes the default).

I personally don't really like it and would prefer to just call a different constructor in favor of better names.

I like the idea of two camera classes and a neatly encapsulated lower level. However, if we take that route, I think we should follow it all the way to the end - and not add properties unrelated to the bare setup of the projection matrix. A projection doesn't focus - only a physical camera is actually capable of that!

We might as well go with the current state in this PR and have just one camera that has all the properties. If memory footprint is a concern, we can move the default for new property (which may rarely be set) to the prototype.

Note that *I didn't actually add anything*, but instead just tied the lose ends together, on the same hand making it more compatible with what is pretty much THE standard rendering interface between authoring tools and renderers.
`PerspectiveCamera` implies a 35mm film, right now: The current implementation of `setLens` takes a second parameter `frameHeight` to override it. It is rather awkward to use, because the user has to apply a fixed aspect as for some photo format and then gets an estimate for the reward of doing the math himself :). This is why this parameter actually defaults to 24 (millimeters of film, assuming the fixed aspect of a landscape photo) and probably why noone ever dared to actually ever use it directly (ironically though, the API is dutifully wrapped by two other example classes). And yeah, the documentation is off, stating it would be the film size in some places.

Now, calling `setFocalLength` or `setLens` (I deprecated that one), you get the real thing. And you can take the current state and easily export it to an external renderer (e.g. for server-side ray tracing).

bhouston commented:
Other data points....

Game engines:
- Unity 3D Camera class: http://docs.unity3d.com/ScriptReference/Camera.html
- Unreal Engine 3D Camera class: https://docs.unrealengine.com/latest/INT/API/Runtime/Engine/Camera/FMinimalViewInfo/index.html

Weird, but I can not find good stereo references for Unity and Unreal Engine.  Argh.

3D editors:
- Softimage: http://softimage.wiki.softimage.com/xsidocs/property1503.htm
- Modo: http://modo.docs.thefoundry.co.uk/modo/501/help/pages/shaderendering/CameraItem.html

The 3D editors all use film size, Modo interestingly has a FilmFit parameter that determines how one fits the screen to the film size, e.g. black bars nor not.  Actually I personally thing Modo's parameters are the most thorough and I like how it groups some parameters under DOF and Stereoscopic categories.

mrdoob commented:
Thanks!
